handler/src/test/java/io/netty/handler/ssl/SniClientJava8TestUtil.java

```diff
@@ -0,0 +1,128 @@
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.DefaultEventLoopGroup;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.local.LocalAddress;
+import io.netty.channel.local.LocalChannel;
+import io.netty.channel.local.LocalServerChannel;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import io.netty.handler.ssl.util.SelfSignedCertificate;
+import io.netty.util.concurrent.Promise;
+import io.netty.util.internal.ThrowableUtil;
+
+import javax.net.ssl.SNIMatcher;
+import javax.net.ssl.SNIServerName;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.SSLParameters;
+import java.util.Collections;
+
+/**
+ * In extra class to be able to run tests with java7 without trying to load classes that not exists in java7.
+ */
+final class SniClientJava8TestUtil {
+
+    private SniClientJava8TestUtil() { }
+
+    static void testSniClient(SslProvider sslClientProvider, SslProvider sslServerProvider, final boolean match)
+            throws Exception {
+        final String sniHost = "sni.netty.io";
+        LocalAddress address = new LocalAddress("test");
+        EventLoopGroup group = new DefaultEventLoopGroup(1);
+        Channel sc = null;
+        Channel cc = null;
+        try {
+            SelfSignedCertificate cert = new SelfSignedCertificate();
+            final SslContext sslServerContext = SslContextBuilder.forServer(cert.key(), cert.cert())
+                    .sslProvider(sslServerProvider).build();
+            final Promise<Void> promise = group.next().newPromise();
+            ServerBootstrap sb = new ServerBootstrap();
+            sc = sb.group(group).channel(LocalServerChannel.class).childHandler(new ChannelInitializer<Channel>() {
+                @Override
+                protected void initChannel(Channel ch) throws Exception {
+                    SslHandler handler = sslServerContext.newHandler(ch.alloc());
+                    SSLParameters parameters = handler.engine().getSSLParameters();
+                    SNIMatcher matcher = new SNIMatcher(0) {
+                        @Override
+                        public boolean matches(SNIServerName sniServerName) {
+                            return match;
+                        }
+                    };
+                    parameters.setSNIMatchers(Collections.singleton(matcher));
+                    handler.engine().setSSLParameters(parameters);
+
+                    ch.pipeline().addFirst(handler);
+                    ch.pipeline().addLast(new ChannelInboundHandlerAdapter() {
+                        @Override
+                        public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+                            if (evt instanceof SslHandshakeCompletionEvent) {
```

Comment:
this looks familiar to me and something very similar is used in SSLEngineTest and SslHandlerTest  ... can we consolidate the duplicated code? followup PR is also fine.